pkg/controller/egress/controller.go

```diff
@@ -43,11 +55,37 @@ const (
 	defaultWorkers = 4
 	// egressGroupType is the type used when registering EgressGroups to the grouping interface.
 	egressGroupType grouping.GroupType = "egressGroup"
+
+	externalIPPoolIndex = "externalIPPool"
+)
+
+var (
+	externalIPPoolNotFound = errors.New("ExternalIPPool not found")
 )
 
+// ipAllocation contains the IP and the IP Pool which allocates it.
+type ipAllocation struct {
+	ip     net.IP
+	ipPool string
+}
+
 // EgressController is responsible for synchronizing the EgressGroups selected by Egresses.
 type EgressController struct {
+	crdClient                  clientset.Interface
+	externalIPPoolLister       egresslisters.ExternalIPPoolLister
+	externalIPPoolListerSynced cache.InformerSynced
+	// ipAllocatorMap is a map from ExternalIPPool name to MultiIPAllocator.
+	ipAllocatorMap   map[string]ipallocator.MultiIPAllocator
+	ipAllocatorMutex sync.RWMutex
```

Comment:
you are not using `ipAllocatorMutex` anywhere, I assume it's because of a copy-paste error of the Lock / Unlock code

Reply:
s/copy-paste error/auto-completion error/ :)